pippy/PipelineSchedule.py

```diff
@@ -507,17 +569,26 @@ def backward_one_chunk(self, **kwargs) -> None:
 
         self.inputs_grad = gradients
 
-    def compute_loss(self):
-        if self.outputs is None:
-            raise RuntimeError("forward() must be called before compute_loss()")
-        # TODO: use a real loss function passed in
-        return self.outputs[0].mean()
-
 
 class PipelineSchedule(ABC):
-    def __init__(self, stage: PipelineStageBase, n_microbatches: int):
+    def __init__(
+        self,
+        stage: PipelineStageBase,
+        n_microbatches: int,
+        loss_fn: Optional[Callable] = None,
```

Comment:
I agree loss_fn should not be an argument for step function. So I am giving it to the constructor of schedule. The main idea being: 
(1) let's leave the stages alone, because loss_fn is not part of the model in this case; thus we should try to make stages loyal to the original model. 
(2) calculating loss is usually in the training step, hence it is put in schedule now. (schedule is a pre-defined pattern for training step).
(3) Of course, if model is defined with loss inside, we should respect that too, as in tracer path's previous implementation. I am dropping this support in this first implementation, but will add it back after revisiting model definitions on HF.

Reply:
I was saying loss_fn could potentially be an argument in the `step()` function.
1) Agree that loss isnt related to the stages / model
2) schedule doesn't necessarily need to be only for training, you could (and we can provide a default) for a schedule with only forward passes for inference? Not sure if that is overkill

Reply:
Re 2:
Right. In the inference case, user will not/need not to pass the `loss_fn` to schedule. `loss_fn` is Optional.
In fact, in the code, I check whether `loss_fn is None` to decide whether this schedule should do backward or not.

Reply:
I am not sure if user would have need to change `loss_fn` between steps. If they really have such need, they can recreate a schedule.